### ISSUE TITLE
WhistelistType: `ALLOWED` vs `ALLOWED_OFFLINE` and `ALWAYS`

### ISSUE THREAD

**Issue by bramus** [ocpi/ocpi#186]
Right now the docs for the `WhistelistType` of a `Token` read this:

| Value                           | Description                                                                |
|---------------------------------|----------------------------------------------------------------------------|
| ALWAYS                          | Token always has to be whitelisted, [realtime authorization](mod_tokens.md#13-real-time-authorization) is not possible/allowed. |
| ALLOWED                         | It is allowed to whitelist the token, [realtime authorization](mod_tokens.md#13-real-time-authorization) is also allowed.   |
| ALLOWED_OFFLINE                 | Whitelisting is only allowed when CPO cannot reach the eMSP (communication between CPO and eMSP is offline)                  |
| NEVER                           | Whitelisting is forbidden, only [realtime authorization](mod_tokens.md#13-real-time-authorization) is allowed. Token should always be authorized by the eMSP. |

The values `ALWAYS` and `NEVER` are straightforward: you either don't or do make an authorization request for said token. If said request should timeout (due to a broken link between the eMSP and CPO), then you disallow the token because it cannot be verified.

`ALLOWED_OFFLINE` is also well described and clear: you make the authorization request
  - If the request timeouts (due to a broken link between the eMSP and CPO) you allow the token
  - If the request doesn't timeout you evaluate the `valid` field from the response

~

However, I'm a bit confused about the `ALLOWED` option. Basically you immediately _(without making any request)_ allow the the token, or you do perform a realtime authorization request. But the result of said request can be interpreted in a few ways:

## Interpretation 1

- If the request timeouts (due to a broken link between the eMSP and CPO) you allow the token because the (first part of its) description reads “It is allowed to whitelist the token”.
- If the request doesn't timeout you evaluate the `valid` field from the response.

**With this interpretation both `ALLOWED_OFFLINE` and `ALLOWED` scenarios yield the same result.**

## Interpretation 2

- If the request timeouts (due to a broken link between the eMSP and CPO) you disallow the token, even though that is in conflict with the first part of its description.
- If the request doesn't timeout you evaluate the `valid` field from the response.

**With this interpretation both `ALLOWED` and `NEVER` scenarios yield the same result: a request is made and its result is evaluated.**

## Interpretation 2, bis

Building further upon interpretation 2 you, as a CPO, opt to not make the realtime authorization request in the first at all, as it would only prevent a user from charging.

**With this interpretation both `ALLOWED` and `ALWAYS` scenarios yield the same result: no request is made and the token is allowed.**

~

So, as explained above the `ALLOWED` scenario can yield the same result as either `ALLOWED_OFFLINE`, `ALWAYS`, or `NEVER`.

Could it be that I overlooking something? One thing that I can think of is that when the value is `ALLOWED`, it actually means that the token _may_ be sent over to the CP so that it can be stored over there in its local whitelist.
- If you, as a CPO, opt to use the local whitelist, the token is immediately allowed upon swiping.
- If you, as a CPO, don't opt to use the local whitelist, you fall back to the realtime authorization request (and continue as in the case of `NEVER`: a timeout will disallow the token).

If that indeed is the case, a rewording of the description for `ALLOWED` is recommended … a mention of CP and/or "local whitelist" seem feasible.

If it's not then … can someone enlighten me? Touching a bit in the dark right now.

**Comment by RobertdeLeeuw-IHomer** [comment]
The OCPI White-listing does not have to have a relation with OCPP. Yes, as CPO, you can send tokens to the Whitelist of a Charge Point. But OCPI does not say anything about this.

'allowed': As MSP I dont care how the CPO does the authorization, it can authorize a transaction themselves (whitelist) or do a realtime authorization. This choose is left to the CPO. (I guess that most CPO will see this as: ALWAYS, and dont do a real-time autorization)

'allowed_offline': "This is: do realtime authorization, unless the MSP cannot be reached, request times out, then the CPO might authorize them selves. (But the CPO is not forced to do so, I the CPO is in doubt that it will get payed by a MSP when they authorize a transaction when: "MSP is offline" they might, if they really want to, reject the authorization. (I Guess nobody would reject when MSP is offline)

**Comment by poohsen** [comment]
yeah, that was my interpretation as well. Both, ALLOWED and ALLOWED_OFFLINE add an additional realtime auth request which is meant to enhance the security. 

In the case of ALLOWED, the decision to take advantage of this lies with the CPO. Whether to go with the behaviour you describe in interpretation 1, 2 or "2, bis" , that's up to the CPO.

ALLOWED_OFFLINE gives the MSP a way to try for more security and hopefully fewer problematic/disputable sessions and the CPO should accomodate this initiative.

Whether the distinction is useful at all I can't say. We hardcode ALLOWED_OFFLINE as an MSP and this fits our processes well. The different options put more burden on CPOs' implementations so it would be interesting to hear if other MSPs take advantage of the other enum values.
